### PR TITLE
DRAFT feat: conditional update poc

### PR DESCRIPTION
**Pull request by djordje-mijatovic**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java [hunk 1]
```diff
@@ -553,10 +559,20 @@ private static GlossaryTermAssociationArray removeTermsIfExists(GlossaryTerms te
...
+    Map<String, Long> createdOnMap = ETagUtil.extractETag(eTag);
     for (MetadataChangeProposal change : changes) {
-      entityService.ingestProposal(change, getAuditStamp(actor), false);
+      if (createdOnMap.containsKey(change.getEntityUrn())) {
+        entityService.ingestProposal(change, getAuditStamp(actor), false, createdOnMap.get(change.getEntityUrn()));
+      } else {
+        entityService.ingestProposal(change, getAuditStamp(actor), false);
+      }
+
+      // ETAG Comment: If - Esle branch from about should be deleted and only if branch kept.
```
Comment (RyanHolstien) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java:572]: Can we mark the eTag as Nullable as we pass it around if this is the intent? So we know that it might be null, and will be for all current use cases, as it goes down the stack.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java [hunk 2]
```diff
@@ -27,6 +27,8 @@ public class AddTagResolver implements DataFetcher<CompletableFuture<Boolean>> {
 
   @Override
   public CompletableFuture<Boolean> get(DataFetchingEnvironment environment) throws Exception {
+    // ETAG Comment: Client should send the eTag as part of the Variables. Here we received it for GraphQL implementation.
+    final String eTag = environment.getVariables().containsKey("eTag") ? environment.getVariables().get("eTag").toString() : null;
```
Comment (RyanHolstien) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java:31]: This seems to imply the ETag is expected to come from the client rather than from the server. Per the RFC, client should be sending in the If-Match header (or perhaps [If-Unmodified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since) makes more sense here) and the server should be generating the ETag to be returned as a response header:

```
We could achieve this via HTTP [Conditional Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#avoiding_the_lost_update_problem_with_optimistic_locking).
GMS would populate the `ETag` header in the GET response for aspects. This can be used in the `If-Match` or 
`If-None-Match` header when a POST request is passed to GMS to update the aspects. The main concern is then how to 
format the ASCII `Etag` value for multiple aspects such that client and server are able to compose them.

A suggested format would be semicolon-separated aspect-value pairs, e.g. `myAspect1=hashValue1;myAspect2=hashValue2`.
```

Clients don't typically send in the ETag header to my understanding as it's a server side identification concept to be used in serial requests dependent on eachother. I.E.:

Client -> Server (ETag: XYZ) -> Client (If-Match: XYZ) -> Server

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java (hunk 1 above)
Comment (djordje-mijatovic) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java:572]: Yes

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java (hunk 2 above)
Comment (djordje-mijatovic) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java:31]: It is doable. After some internal discussion we had about this, and also about condition insert, here are some conclusions and suggestions for the next steps:
1) Instead of using eTag, we should use "If-Unmodified-Since" header. I somehow think this one is better choice than "If-Match".
2) We should use the same header for conditional insert. If "If-Unmodified-Since" is 0 than we do the insert only if the aspect doesn't exist. We could introduce another header for this but I would avoid it as it is simpler to have one variable.

To sum up, here are the possible values (and meanings) for "If-Unmodified-Since" header:
1) If "If-Unmodified-Since" is greater then 0 we do the conditional update. 
2) If "If-Unmodified-Since" is equals 0 we do the conditional insert. I see "If-Unmodified-Since = 0" as "If Unmodified Since the beginning of the time == If not Created Yet"
3) If "If-Unmodified-Since" is null we do NOT apply any conditional operation.

For GraphQL I would suggest to use variable with the same name "If-Unmodified-Since" to provide the information.

Comment (RyanHolstien) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java:31]: Yep having 0 have meaning makes sense to me, this is not a default header that gets added so if someone is adding this header in it is expected to actually mean something. Using If-Unmodified-Since makes more sense to me as well since we're using a timestamp driven approach.

Comment (mattmatravers) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/AddTagResolver.java:31]: @RyanHolstien, note that `If-Unmodified-Since == 0` is effectively the `CREATE` [ChangeType](https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/events/metadata/ChangeType.pdl). I'd suggest a follow-up PR later on that could wire this together if there is interest.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java [hunk 3]
```diff
@@ -559,9 +563,12 @@ private SimpleStatement generateSaveStatement(EntityAspect aspect, boolean inser
 
       Update u = uwa.whereColumn(CassandraAspect.URN_COLUMN).isEqualTo(literal(aspect.getUrn()))
               .whereColumn(CassandraAspect.ASPECT_COLUMN).isEqualTo(literal(aspect.getAspect()))
-              .whereColumn(CassandraAspect.VERSION_COLUMN).isEqualTo(literal(aspect.getVersion()))
-              .ifExists();
-
+              .whereColumn(CassandraAspect.VERSION_COLUMN).isEqualTo(literal(aspect.getVersion()));
+      if (updateIfCreatedOn != null) {
+        u = u.ifRaw(CassandraAspect.CREATED_ON_COLUMN + "=" + updateIfCreatedOn);
```
Comment (justinas-marozas) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:568]:
```suggestion
        u = u.whereColumn(CassandraAspect.CREATED_ON_COLUMN).isEqualTo(literal(updateIfCreatedOn));
```

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java [hunk 4]
```diff
@@ -553,10 +561,11 @@ private static GlossaryTermAssociationArray removeTermsIfExists(GlossaryTerms te
     return termAssociationArray;
   }
 
-  private static void ingestChangeProposals(List<MetadataChangeProposal> changes, EntityService entityService, Urn actor) {
+  private static void ingestChangeProposals(List<MetadataChangeProposal> changes, EntityService entityService, Urn actor, String condUpdate) {
     // TODO: Replace this with a batch ingest proposals endpoint.
+    Map<String, Long> createdOnMap = CondUpdateUtils.extractCondUpdate(condUpdate);
     for (MetadataChangeProposal change : changes) {
-      entityService.ingestProposal(change, getAuditStamp(actor), false);
+      entityService.ingestProposal(change, getAuditStamp(actor), false, createdOnMap.get(change.getEntityUrn()));
```
Comment (justinas-marozas) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/LabelUtils.java:568]: I suspect `createdOnMap.get(change.getEntityUrn())` will always return `null`. Keys in `createOnMap` are String but `change.getEntityUrn()` returns a Urn. Looking at `equals` overrides, I doubt a Urn can ever be equal to a String.

File: metadata-service/auth-impl/src/main/java/com/datahub/authentication/token/StatefulTokenService.java [hunk 5]
```diff
@@ -129,8 +130,9 @@ public String generateAccessToken(@Nonnull final TokenType type, @Nonnull final
     // Need this to write key aspect
     final List<MetadataChangeProposal> additionalChanges = AspectUtils.getAdditionalChanges(proposal, _entityService);
 
-    _entityService.ingestProposal(proposal, auditStamp, false);
-    additionalChanges.forEach(mcp -> _entityService.ingestProposal(mcp, auditStamp, false));
+    Map<String, Long> createdOnMap = CondUpdateUtils.extractCondUpdate(condUpdate);
+    _entityService.ingestProposal(proposal, auditStamp, false, createdOnMap.get(proposal.getEntityUrn()));
+    additionalChanges.forEach(mcp -> _entityService.ingestProposal(mcp, auditStamp, false, createdOnMap.get(proposal.getEntityUrn())));
```
Comment (justinas-marozas) [inline metadata-service/auth-impl/src/main/java/com/datahub/authentication/token/StatefulTokenService.java:135]: Same as LabelUtils.

**Comment by aditya-radhakrishnan** [comment]
Hey there! Since this is marked as a Draft, I'm going to change the status from Open to Draft. Please feel free to change it back when it's ready!
